### PR TITLE
For check expression, take error from the expectedType when possible

### PR DESCRIPTION
**Pull request by rdhananjaya**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/30737

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -5135,6 +5135,8 @@ private void visitCheckAndCheckPanicExpr(BLangCheckedExpr checkedExpr) {
         BType typeOfExprWithCheckingKeyword;
         if (expType == symTable.noType) {
             typeOfExprWithCheckingKeyword = symTable.noType;
+        } else if (expType.tag == TypeTags.UNION && types.getErrorType((BUnionType) expType) != null) {
+            typeOfExprWithCheckingKeyword = expType;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5139]:
```ballerina
type Err distinct error;

type ErrTwo distinct error;

public function main() returns Err? {
    int|ErrTwo x = check bar();
}

function bar() returns int|error {
    return error("");
}
```

This now results in the following error, which isn't quite correct, right?
```cmd
incompatible types: expected '(int|ErrTwo)', found '(int|error)'
```

Previously the error was 
```cmd
invalid usage of the 'check' expression operator: no matching error return type(s) in the enclosing invokable
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/checkedexpr/CheckedExpressionOperatorTest.java [hunk 2]
```diff
@@ -228,6 +228,11 @@ public void testCheckedErrorsWithReadOnlyInUnion() {
         BRunUtil.invoke(result, "testCheckedErrorsWithReadOnlyInUnion");
     }
 
+    @Test
+    public void testCheckWithTrinaryOperator() {
```
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/checkedexpr/CheckedExpressionOperatorTest.java:232]: My bad, let me fix

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 1 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: This is fixed now. Thank you.

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/checkedexpr/checked_error_return_type_mismatch_negative.bal [hunk 3]
```diff
@@ -47,3 +47,14 @@ public function baz() {
...
+type LhsErr distinct error;
+type LhsErrTwo distinct error;
+
+public function fooBar() returns LhsErr? {
+    int|LhsErrTwo x = check fbar();
+}
+
+function fbar() returns int|error {
+    return error("");
+}
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/checkedexpr/checked_error_return_type_mismatch_negative.bal:60]: Should we also test something like

```ballerina
public function test() returns LhsErr? {
    int|LhsErr x = check func();
    int y = check func();
    int|LhsErrTwo z = check func();
}

function func() returns LhsErrTwo|int {
    return error("");
}
```

to verify the type ID check in the same type check?

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/checkedexpr/checked_expr_operator_basics.bal [hunk 4]
```diff
@@ -274,6 +274,28 @@ function testCallExprWithCheck() {
...
+function barOrErr() returns string|Err {
+    return "bar";
+}
+
+function bazOrErr() returns string|Err {
+    return error("Err");
+}
+
+function testCheckWithTernaryOperator() {
+    var a = ternaryCheck(false);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/checkedexpr/checked_expr_operator_basics.bal:292]: Shall we add a test for when a string is returned also?

**Review by MaryamZi — APPROVED** [review]

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**